File: bitcoin_stockmem/data/price_fetcher.py
"""Fetch daily OHLCV data for BTC and ETH from Binance via ccxt."""

from __future__ import annotations

import logging
from datetime import datetime, timedelta

import ccxt
import pandas as pd

from config import TRADING_PAIRS

logger = logging.getLogger(__name__)


def fetch_daily_ohlcv(
    asset: str = "BTC",
    start_date: str = "2025-01-01",
    end_date: str | None = None,
) -> pd.DataFrame:
    """Fetch daily candles from Binance.

    Args:
        asset: "BTC" or "ETH".
        start_date: ISO date string (YYYY-MM-DD).
        end_date: ISO date string; defaults to today.

    Returns:
        DataFrame with columns: date, open, high, low, close, volume, return_pct
    """
    exchange = ccxt.binance({"enableRateLimit": True})
    symbol = TRADING_PAIRS[asset]

    since = exchange.parse8601(f"{start_date}T00:00:00Z")
    end_dt = datetime.fromisoformat(end_date) if end_date else datetime.utcnow()
    end_ts = int(end_dt.timestamp() * 1000)

    all_candles = []
    while since < end_ts:
        candles = exchange.fetch_ohlcv(symbol, "1d", since=since, limit=500)
        if not candles:
            break
        all_candles.extend(candles)
        since = candles[-1][0] + 86_400_000  # next day
        if len(candles) < 500:
            break

    df = pd.DataFrame(
        all_candles,
        columns=["timestamp", "open", "high", "low", "close", "volume"],
    )
    df["date"] = pd.to_datetime(df["timestamp"], unit="ms").dt.strftime("%Y-%m-%d")
    df = df[df["date"] <= (end_date or end_dt.strftime("%Y-%m-%d"))]
    df = df.drop_duplicates(subset="date", keep="last").sort_values("date")

    # Daily return
    df["return_pct"] = df["close"].pct_change()
    df = df.reset_index(drop=True)

    logger.info(f"Fetched {len(df)} daily candles for {asset}")
    return df[["date", "open", "high", "low", "close", "volume", "return_pct"]]

File: bitcoin-stockmem-ts/src/data/price-fetcher.ts
/**
 * Fetch daily OHLCV data for BTC and ETH from Binance via ccxt.
 */

import ccxt from "ccxt";
import { TRADING_PAIRS, type Asset } from "../config";

export interface PriceRow {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  return_pct: number | null;
}

export async function fetchDailyOhlcv(
  asset: Asset = "BTC",
  startDate: string = "2025-01-01",
  endDate?: string
): Promise<PriceRow[]> {
  const exchange = new ccxt.binance({ enableRateLimit: true });
  const symbol = TRADING_PAIRS[asset];

  let since = exchange.parse8601(`${startDate}T00:00:00Z`)!;
  const endDt = endDate ? new Date(endDate) : new Date();
  const endTs = endDt.getTime();

  const allCandles: number[][] = [];
  while (since < endTs) {
    const candles = await exchange.fetchOHLCV(symbol, "1d", since, 500);
    if (!candles.length) break;
    for (const c of candles) {
      allCandles.push(c.map((v) => v ?? 0));
    }
    since = (candles[candles.length - 1][0] ?? 0) + 86_400_000;
    if (candles.length < 500) break;
  }

  const endStr = endDate || endDt.toISOString().slice(0, 10);
  const rows: PriceRow[] = [];
  const seen = new Set<string>();

  for (const c of allCandles) {
    const date = new Date(c[0]).toISOString().slice(0, 10);
    if (date > endStr || seen.has(date)) continue;
    seen.add(date);
    rows.push({
      date,
      open: c[1],
      high: c[2],
      low: c[3],
      close: c[4],
      volume: c[5],
      return_pct: null,
    });
  }

  rows.sort((a, b) => a.date.localeCompare(b.date));

  // Compute daily returns
  for (let i = 1; i < rows.length; i++) {
    rows[i].return_pct = (rows[i].close - rows[i - 1].close) / rows[i - 1].close;
  }

  console.log(`Fetched ${rows.length} daily candles for ${asset}`);
  return rows;
}
